### PR TITLE
Add Scotch v7.0.1, now supporting CMake

### PR DESCRIPTION
**Pull request by pghysels**
I'm not sure if/when to enable/disable the MPI_THREAD_MULTIPLE and THREADS options in Scotch's CMake.

For v7.0.1 I do not see a compression option in the CMake installer.

### CODE REVIEW

File: var/spack/repos/builtin/packages/scotch/package.py [hunk 1]
```diff
@@ -262,3 +270,33 @@ def install(self, spec, prefix):
...
+            self.define_from_variant('INSTALL_METIS_HEADERS', 'metis'),
+            self.define_from_variant('BUILD_LIBESMUMPS', 'esmumps'),
+            self.define_from_variant('BUILD_SHARED_LIBS', 'shared')
+        ]
+
+        if '+mpi' in spec:
+            args.append('-DBUILD_PTSCOTCH=ON')
+            # TODO check if MPI_THREAD_MULTIPLE is supported?
+        else:
+            args.append('-DBUILD_PTSCOTCH=OFF')
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/scotch/package.py]:
```suggestion
       args.append(self.define_from_variant('BUILD_PTSCOTCH', 'mpi')
```

**Review by sethrj — COMMENTED** [review]
Yay cmake!

File: var/spack/repos/builtin/packages/scotch/package.py (hunk 1 above)
Comment (pghysels) [reply var/spack/repos/builtin/packages/scotch/package.py]: I was going to enable/disable `MPI_THREAD_MULTIPLE` in the if branch. But I don't see a simple way to check if it is supported.

File: var/spack/repos/builtin/packages/scotch/package.py [hunk 2]
```diff
@@ -95,13 +100,15 @@ def libs(self):
 
         return scotchlibs + zlibs
 
+    @when('@:6.9.99')
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/scotch/package.py]:
```suggestion
    @when('@:6')
```
Here and below.

File: var/spack/repos/builtin/packages/scotch/package.py [hunk 3]
```diff
@@ -6,15 +6,20 @@
 from spack import *
 
 
-class Scotch(Package):
+class Scotch(CMakePackage):
```
Comment (sethrj) [inline var/spack/repos/builtin/packages/scotch/package.py:9]: Unfortunately this makes `cmake` a dependency even for older versions that don't use it... I think that's a fair tradeoff though.

**Review by sethrj — COMMENTED** [review]
One more suggestion. I tried building an older version (6.1.3) on my mac (apple-clang%13.0.0) but hit errors due to zealous warning flags:
```
 common.c:116:3: error: implicit declaration of function 'gettimeofday' is invalid in
           C99 [-Werror,-Wimplicit-function-declaration]
```
and even on 7.0.1 it looks like the library does not correctly declare its dependencies:
```
 >> 561    Undefined symbols for architecture arm64:
     562      "_SCOTCH_errorPrint", referenced from:
     563          __SCOTCHarchLoad in arch.c.o
     564          __SCOTCHarchSave in arch.c.o
     565          __SCOTCHarchDecoArchBuild in arch_build.c.o
     566          __SCOTCHarchDeco2ArchBuild in arch_build2.c.o
     567          _archDeco2BuildMatchMate in arch_build2.c.o
     568          __SCOTCHarchCmpltArchLoad in arch_cmplt.c.o
     569          __SCOTCHarchCmpltArchSave in arch_cmplt.c.o
     570          ...
     571      "_SCOTCH_errorPrintW", referenced from:
     572          _scotchyyparse in parser_yy.c.o
     573    ld: symbol(s) not found for architecture arm64
```
but these are separate issues.

File: var/spack/repos/builtin/packages/scotch/package.py (hunk 3 above)
Comment (pghysels) [reply var/spack/repos/builtin/packages/scotch/package.py:9]: Looking forward to one day removing the old stuff

**Comment by pghysels** [comment]
I tried with clang@13.0.0 on Ubuntu 21.10:
`spack install scotch~mpi %clang@13.0.0`
`spack install scotch~mpi@6.1.3 %clang@13.0.0`
and both installed fine. Are your issues related to arm, or mac?
I will look into it further.

**Review by sethrj — APPROVED** [review]

**Comment by sethrj** [comment]
@pghysels The apple clang compilers don't exactly mirror the official LLVM clang in terms of capabilities, default warnings, etc. -- which may explain the 6.1.3 failure. But it's generally best for compatibility to avoid adding `-Werror` to release builds because you can't tell what future compilers might label a warning.

The v7 failure is because macOS by default requires all shared library symbols to be resolved, whereas the default linker on most linux distributions does not. I think you could replicate the same behavior if you add `-Wl,-z,defs` to your `CMAKE_SHARED_LINKER_FLAGS`.

I've merged because the build errors are separate issues with Scotch rather than a problem with the changeset here.

**Pull request merged**